riscv/runtime/src/fmt.rs

```diff
@@ -58,14 +78,26 @@ mod buf_formatter {
 
     impl<'a> fmt::Write for BufFormatter<'a> {
         fn write_str(&mut self, s: &str) -> fmt::Result {
-            let raw_s = s.as_bytes();
-            let write_len = raw_s.len();
-            if write_len > self.buffer.len() - self.used {
-                return Err(fmt::Error);
+            if self.truncated {
+                return Ok(());
             }
-            let s_uninit: &[MaybeUninit<u8>] = unsafe { core::mem::transmute(&raw_s[..write_len]) };
-            self.buffer[self.used..][..write_len].copy_from_slice(s_uninit);
+
+            let remaining_bytes = self.buffer.len() - self.used;
+            let write_len = s.floor_char_boundary(remaining_bytes);
+            let raw_s = s.as_bytes();
+
+            MaybeUninit::write_slice(
+                &mut self.buffer[self.used..][..write_len],
+                &raw_s[..write_len],
+            );
+
             self.used += write_len;
+
+            if write_len != raw_s.len() {
+                // Mark that we have truncated a string, so we stop writing even
```

Comment:
Maybe this whole class should be renamed to conform with these changes, but this is used it in the `print_truncate!` flag, that is now supposed to truncate and not fail if the buffer is too small. That is because most people don't expect `print`-family macros to panic.

Reply:
Which I also use in the panic handler now, so there is one less chance of panic handler panicking.

Reply:
I really think we should flag an error if we don't perform the full write. We can make the buffer a megabyte if that helps.